### ISSUE TITLE
Move Mailer to SwiftMailer

### ISSUE THREAD

**Issue by silverstripe-issues** [silverstripe/silverstripe-framework#1366]
created by: @chillu (ischommer)
assigned to: @chillu (ischommer)
created at: 2009-01-25
original ticket: http://open.silverstripe.org/ticket/3427

---

Currently we have our own (procedural SilverStripe 1) code for sending html and plaintext emails. It contains all kinds of funky exceptions and tweaks which are not tested nor understood by many ;) There's also a lot of redundancy (and slight differences) between the plaintext and html implementations. We also see increased demand in sending unicode emails in all kinds of flavours, which increases the scenarios which we most likely haven't tested in full before.

I'd suggest we move to a solid, documented and unit tested Mailer implementation.
Originally I've suggested PHPMailer (http://phpmailer.codeworxtech.com), but its license incompatible (GPL).
Main candidate is Zend_Mail - http://framework.zend.com/manual/en/zend.mail.html (would fit well with our other use of Zend).

The main task here is to identify which features overlap and what we need to re-implement - which I expect is very little.

We should make the base mailer implementation pluggable, although it might not be feasible to support all potential options through a wrapper (e.g. adding attachments).

Some examples of the weirdness that is the current implementation:

```
$subjectIsUnicode = (strpos($subject,"&#") !== false);
```

```
function QuotedPrintable_encode($quotprint) {       
        $quotprint = (string) str_replace('\r\n',chr(13).chr(10),$quotprint);
        $quotprint = (string) str_replace('\n',  chr(13).chr(10),$quotprint);
        $quotprint = (string) preg_replace("~([\x01-\x1F\x3D\x7F-\xFF])~e", "sprintf('=%02X', ord('\\1'))", $quotprint);
        //$quotprint = (string) str_replace('\=0D=0A',"=0D=0A",$quotprint);
        $quotprint = (string) str_replace('=0D=0A',"\n",$quotprint);    
        $quotprint = (string) str_replace('=0A=0D',"\n",$quotprint);    
        $quotprint = (string) str_replace('=0D',"\n",$quotprint);   
        $quotprint = (string) str_replace('=0A',"\n",$quotprint);   
        return (string) $quotprint;
}
```

**Comment by tractorcow** [comment]
It would be good to do https://github.com/silverstripe/silverstripe-framework/issues/3057 at the same time

**Comment by dhensby** [comment]
I'm kind of working on this https://github.com/dhensby/silverstripe-framework/tree/pulls/4/mailer

**Comment by tractorcow** [comment]
I'm also in favour of swiftmailer... good job @dhensby !

**Comment by patricknelson** [comment]
Would like to see this in 4.0. I've wrestled a bit with the mailer a few times myself and this would be helpful @dhensby. Is anyone aware of any major advantages of SwiftMailer over Zend_Mail or vice versa? e.g. security, activity, general features and etc. I know we already use Zend components elsewhere in SS.

**Comment by tractorcow** [comment]
Related issue: https://github.com/silverstripe/silverstripe-framework/issues/5719

**Comment by sminnee** [comment]
Pull request available in #6466

**Comment by sminnee** [comment]
Merged!
